### PR TITLE
TIP-703: Add validation on the product identifier

### PR DESCRIPTION
**Pull request by damien-carcel**
### Context

We previously introduced a new column for the product identifier. But we never validated it. ThPR fixes this behavior

There are still some weirdnesses about the update of the identifier. Who's responsible for the change? The value or the field? Should we keep the value which seems to cause a lot of problems now? We'll tackle this problem in another card (see PIM-6123 for instance) and we need to discuss that topic with the POs.

### Expected

- Identifier column has been added to the validation of the product
- An integration test ensures we can not create several products with the same identifier
- The problem between value and field is **NOT** resolved here
- The problem about validating several identifiers among a batch is **NOT** resolved here

## Definition Of Done

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | ❎
| Added Behats                      | ❎
| Added integration tests           | ✅
| Changelog updated                 | ✅
| Review and 2 GTM                  | ✅
| Micro Demo to the PO (Story only) | ❎
| Migration script                  | ❎
| Tech Doc                          | ❎

🕐 Pending / Work in progress
✅ Done / Validated
❎ Not needed

### CODE REVIEW

File: features/product/create_product.feature [hunk 1]
```diff
@@ -38,7 +38,7 @@ Feature: Product creation
     And I fill in the following information in the popin:
       | SKU | sandals |
     And I press the "Save" button in the popin
-    Then I should see validation error "The value sandals is already set on another product for the unique attribute sku"
+    Then I should see validation error "This value is already used."
```
Comment (jjanvier) [inline features/product/create_product.feature]: no way to have something more specific? that's Symfony's default?

**Review by jjanvier — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml [hunk 2]
```diff
@@ -1,5 +1,8 @@
 Pim\Component\Catalog\Model\Product:
     constraints:
+        - Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity:
+            fields: [identifier]
+            message: The same identifier is already set on another product
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml]: Shouldn't this be a translation code ?

**Review by SamirBoulil — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml [hunk 3]
```diff
@@ -12,6 +15,14 @@ Pim\Component\Catalog\Model\Product:
             groups:
                 - Default
                 - pim_catalog_variant_group
+    properties:
+        identifier:
+            - NotBlank: ~
+            - Regex:
+                pattern: '/^((?![,;]).)*$/'
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml]: Why not just `^[^,;]+$`?

File: src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        $this->productBuilder = $this->get('pim_catalog.builder.product');
+        $this->productValidator = $this->get('pim_catalog.validator.product');
+        $this->productSaver = $this->get('pim_catalog.saver.product');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getConfiguration()
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php]: protected should be at the end of the test

File: src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    protected $productBuilder;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->productBuilder = $this->get('pim_catalog.builder.product');
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php]: Be careful when making your test case stateful, as we don't know exactly how phpunit works we could have surprises (such as memory leaks). Maybe just use protected methods? (`createProduct`, `validateProduct`...)

File: src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            true
+        );
+    }
+
+    public function testUniqueIdentifierValidation()
+    {
+        $product1 = $this->productBuilder->createProduct('just_an_empty_product');
+        $violations = $this->productValidator->validate($product1);
+
+        $this->assertEquals($violations->count(), 0);
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php]: There is an `assertCount()` if you want.

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductViolationNormalizer.php [hunk 7]
```diff
@@ -44,12 +56,19 @@ public function normalize($violation, $format = null, array $context = [])
                 'attribute' => $productValue->getAttribute()->getCode(),
                 'locale'    => $productValue->getLocale(),
                 'scope'     => $productValue->getScope(),
-                'message'   => $violation->getMessage()
+                'message'   => $violation->getMessage(),
+            ];
+        } elseif ('identifier' === $path) {
+            $normalizedViolation = [
+                'attribute' => $this->attributeRepository->getIdentifierCode(),
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductViolationNormalizer.php:63]: I thought the identifier wasn't related to an attribute anymore... That's confusing.

File: src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml (hunk 3 above)
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml]: I tried but is did not work (but I am kind of a noob in regex, so maybe I didn't write it right).

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductViolationNormalizer.php (hunk 7 above)
Comment (damien-carcel) [reply src/Pim/Bundle/EnrichBundle/Normalizer/ProductViolationNormalizer.php:63]: Yes, that's confusing and need to be addressed later.
The fact is there is still an identifier (and only one) attribute, like it is in the current version, and a corresponding product value. But this value is not saved in the "rawValues" column, but in the "identifier" column.

File: src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php (hunk 4 above)
Comment (damien-carcel) [reply src/Pim/Bundle/CatalogBundle/tests/integration/ProductValidation/UniqueIdentifierValidationIntegration.php]: done.

File: src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml (hunk 3 above)
Comment (damien-carcel) [reply src/Pim/Bundle/CatalogBundle/Resources/config/validation/product.yml]: Changed

**Review by phaseinducer — APPROVED** [review]

**Pull request merged**